test/integration/cds_integration_test.cc

```diff
@@ -0,0 +1,211 @@
+                                               discovery_request.error_detail().code(),
+                                               expected_error_code);
+    }
+    EXPECT_TRUE(
+        IsSubstring("", "", expected_error_message, discovery_request.error_detail().message()));
+    const std::vector<std::string> resource_names(discovery_request.resource_names().cbegin(),
+                                                  discovery_request.resource_names().cend());
+    if (expected_resource_names != resource_names) {
+      return AssertionFailure() << fmt::format(
+                 "resources {} do not match expected {} in {}",
+                 fmt::join(resource_names.begin(), resource_names.end(), ","),
+                 fmt::join(expected_resource_names.begin(), expected_resource_names.end(), ","),
+                 discovery_request.DebugString());
+    }
+    // TODO(PiotrSikora): Remove this hack once fixed internally.
+    if (!(expected_version == discovery_request.version_info())) {
+      return AssertionFailure() << fmt::format("version {} does not match expected {} in {}",
+                                               discovery_request.version_info(), expected_version,
+                                               discovery_request.DebugString());
+    }
+    return AssertionSuccess();
+  }
+
+  template <class T>
+  void sendDiscoveryResponse(const std::string& type_url, const std::vector<T>& messages,
+                             const std::string& version) {
+    envoy::api::v2::DiscoveryResponse discovery_response;
+    discovery_response.set_version_info(version);
+    discovery_response.set_type_url(type_url);
+    for (const auto& message : messages) {
+      discovery_response.add_resources()->PackFrom(message);
+    }
+    cds_stream_->sendGrpcMessage(discovery_response);
+  }
+
+  envoy::api::v2::Cluster buildCluster(const std::string& name) {
+    return TestUtility::parseYaml<envoy::api::v2::Cluster>(
+        fmt::format(R"EOF(
+      name: {}
+      connect_timeout: 5s
+      type: STATIC
+      load_assignment:
+        cluster_name: {}
+        endpoints:
+        - lb_endpoints:
+          - endpoint:
+              address:
+                socket_address:
+                  address: {}
+                  port_value: {}
+      lb_policy: ROUND_ROBIN
+      http2_protocol_options: {{}}
+    )EOF",
+                    name, name, Network::Test::getLoopbackAddressString(ipVersion()),
+                    fake_upstreams_[0]->localAddress()->ip()->port()));
+  }
+
+  void initializeCds() {
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+      // Setup cds and corresponding gRPC cluster.
+      auto* cds_config =
+          bootstrap.mutable_dynamic_resources()->mutable_cds_config()->mutable_api_config_source();
+      bootstrap.mutable_dynamic_resources()->clear_ads_config();
+      bootstrap.mutable_dynamic_resources()->clear_lds_config();
```

Comment:
Do these `clear_*ds_config()` need to happen? I would expect they are empty already.

Reply:
Correct, done.